File: materiales/generate-pdf.py
#!/usr/bin/env python3
"""
Script para generar PDFs bonitos desde archivos Markdown
Usa subprocess para llamar a playwright desde Node.js
"""

import os
import re
import sys
import subprocess
import tempfile
from pathlib import Path

# Plantilla HTML con estilos CSS incluidos
HTML_TEMPLATE = """<!DOCTYPE html>
<html lang="es">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>{title}</title>
    <style>
{css}
    </style>
</head>
<body>
{content}
</body>
</html>
"""

def markdown_to_html(markdown_text):
    """Convierte markdown a HTML con formato básico pero funcional"""
    html = markdown_text

    # Escapar caracteres HTML especiales primero
    # html = html.replace('&', '&amp;').replace('<', '&lt;').replace('>', '&gt;')

    # Headers (de mayor a menor para evitar conflictos)
    html = re.sub(r'^#### (.*?)$', r'<h4>\1</h4>', html, flags=re.MULTILINE)
    html = re.sub(r'^### (.*?)$', r'<h3>\1</h3>', html, flags=re.MULTILINE)
    html = re.sub(r'^## (.*?)$', r'<h2>\1</h2>', html, flags=re.MULTILINE)
    html = re.sub(r'^# (.*?)$', r'<h1>\1</h1>', html, flags=re.MULTILINE)

    # Bold
    html = re.sub(r'\*\*(.*?)\*\*', r'<strong>\1</strong>', html)

    # Italic
    html = re.sub(r'(?<!\*)\*(?!\*)([^*]+)\*(?!\*)', r'<em>\1</em>', html)

    # Lists - primero identificar items de lista
    html = re.sub(r'^- (.*?)$', r'<li>\1</li>', html, flags=re.MULTILINE)

    # Envolver listas consecutivas en <ul>
    lines = html.split('\n')
    in_list = False
    result = []

    for line in lines:
        if line.strip().startswith('<li>'):
            if not in_list:
                result.append('<ul>')
                in_list = True
            result.append(line)
        else:
            if in_list:
                result.append('</ul>')
                in_list = False
            result.append(line)

    if in_list:
        result.append('</ul>')

    html = '\n'.join(result)

    # Horizontal rules
    html = re.sub(r'^---$', '<hr>', html, flags=re.MULTILINE)

    # Párrafos - dividir por líneas vacías
    paragraphs = html.split('\n\n')
    processed = []

    for para in paragraphs:
        para = para.strip()
        if not para:
            continue

        # Si ya tiene una etiqueta HTML, no envolver en <p>
        if (para.startswith('<h') or para.startswith('<ul') or
            para.startswith('<ol') or para.startswith('<hr') or
            para.startswith('<div')):
            processed.append(para)
        else:
            # Reemplazar saltos de línea simples con <br>
            para = para.replace('\n', '<br>\n')
            if not para.startswith('<li>'):
                processed.append(f'<p>{para}</p>')
            else:
                processed.append(para)

    html = '\n\n'.join(processed)

    return html

def generate_html(markdown_file, css_file):
    """Genera HTML desde markdown con estilos CSS"""
    # Leer archivos
    with open(markdown_file, 'r', encoding='utf-8') as f:
        markdown_text = f.read()

    with open(css_file, 'r', encoding='utf-8') as f:
        css_text = f.read()

    # Convertir markdown a HTML
    html_content = markdown_to_html(markdown_text)

    # Obtener título del archivo
    title = Path(markdown_file).stem.replace('_', ' ').title()

    # Crear HTML completo
    full_html = HTML_TEMPLATE.format(
        title=title,
        css=css_text,
        content=html_content
    )

    return full_html

def html_to_pdf_playwright(html_content, output_pdf):
    """Convierte HTML a PDF usando Playwright vía Node.js"""
    # Crear un archivo temporal con el HTML
    with tempfile.NamedTemporaryFile(mode='w', suffix='.html', delete=False, encoding='utf-8') as f:
        temp_html = f.name
        f.write(html_content)

    # Script de Node.js para generar PDF
    node_script = f"""
const {{ chromium }} = require('/opt/node22/lib/node_modules/playwright');
const fs = require('fs');

(async () => {{
    const browser = await chromium.launch();
    const page = await browser.newPage();

    const html = fs.readFileSync('{temp_html}', 'utf-8');
    await page.setContent(html, {{ waitUntil: 'networkidle' }});

    await page.pdf({{
        path: '{output_pdf}',
        format: 'A4',
        margin: {{
            top: '2cm',
            right: '2cm',
            bottom: '2cm',
            left: '2cm'
        }},
        printBackground: true,
        displayHeaderFooter: true,
        headerTemplate: `
            <div style="font-size: 9pt; color: #999; width: 100%; text-align: right; padding-right: 1cm;">
                Curso Intensivo de Español - CILE
            </div>
        `,
        footerTemplate: `
            <div style="font-size: 9pt; color: #999; width: 100%; text-align: center;">
                Página <span class="pageNumber"></span> de <span class="totalPages"></span>
            </div>
        `
    }});

    await browser.close();
}})();
"""

    # Crear archivo temporal con el script de Node.js
    with tempfile.NamedTemporaryFile(mode='w', suffix='.js', delete=False) as f:
        temp_script = f.name
        f.write(node_script)

    try:
        # Ejecutar el script de Node.js
        result = subprocess.run(
            ['node', temp_script],
            capture_output=True,
            text=True,
            timeout=60
        )

        if result.returncode != 0:
            print(f"❌ Error al generar PDF: {result.stderr}")
            return False

        return True

    finally:
        # Limpiar archivos temporales
        try:
            os.unlink(temp_html)
            os.unlink(temp_script)
        except:
            pass

def main():
    """Función principal"""
    script_dir = Path(__file__).parent
    pdf_dir = script_dir / 'pdfs'
    css_file = script_dir / 'pdf-style.css'

    # Crear carpeta para PDFs
    pdf_dir.mkdir(exist_ok=True)

    # Buscar archivos markdown
    md_files = sorted(script_dir.glob('*.md'))

    if not md_files:
        print("❌ No se encontraron archivos .md")
        return

    print(f"\n🚀 Generando PDFs para {len(md_files)} archivos...\n")

    success_count = 0

    for md_file in md_files:
        print(f"📄 Procesando: {md_file.name}")

        try:
            # Generar HTML
            html_content = generate_html(md_file, css_file)

            # Nombre del PDF de salida
            pdf_file = pdf_dir / f"{md_file.stem}.pdf"

            # Convertir a PDF
            if html_to_pdf_playwright(html_content, str(pdf_file)):
                print(f"✅ PDF generado: {pdf_file.name}")
                success_count += 1
            else:
                print(f"❌ Error al generar {pdf_file.name}")

        except Exception as e:
            print(f"❌ Error procesando {md_file.name}: {str(e)}")

    print(f"\n✨ ¡Proceso completado! {success_count}/{len(md_files)} PDFs generados\n")

if __name__ == '__main__':
    main()

File: materiales/generate-pdf.js
#!/usr/bin/env node

/**
 * Script para generar PDFs bonitos desde archivos Markdown
 * Usa Playwright para convertir HTML a PDF
 */

const fs = require('fs');
const path = require('path');
const { chromium } = require('playwright');

// Simple markdown to HTML converter (básico pero funcional)
function markdownToHtml(markdown) {
    let html = markdown;

    // Headers
    html = html.replace(/^# (.*$)/gim, '<h1>$1</h1>');
    html = html.replace(/^## (.*$)/gim, '<h2>$1</h2>');
    html = html.replace(/^### (.*$)/gim, '<h3>$1</h3>');
    html = html.replace(/^#### (.*$)/gim, '<h4>$1</h4>');

    // Bold
    html = html.replace(/\*\*(.*?)\*\*/g, '<strong>$1</strong>');

    // Italic
    html = html.replace(/\*(.*?)\*/g, '<em>$1</em>');

    // Lists
    html = html.replace(/^\- (.*$)/gim, '<li>$1</li>');

    // Wrap lists
    html = html.replace(/(<li>.*<\/li>)/s, function(match) {
        return '<ul>' + match + '</ul>';
    });

    // Line breaks and paragraphs
    html = html.split('\n\n').map(para => {
        if (para.trim().startsWith('<h') ||
            para.trim().startsWith('<ul') ||
            para.trim().startsWith('<ol') ||
            para.trim().startsWith('<li') ||
            para.trim() === '---') {
            return para;
        }
        if (para.trim().length > 0 && !para.trim().startsWith('<')) {
            return '<p>' + para.replace(/\n/g, '<br>') + '</p>';
        }
        return para;
    }).join('\n');

    // Horizontal rules
    html = html.replace(/^---$/gim, '<hr>');

    return html;
}

function createHtmlTemplate(title, content, cssPath) {
    const css = fs.readFileSync(cssPath, 'utf-8');

    return `<!DOCTYPE html>
<html lang="es">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>${title}</title>
    <style>${css}</style>
</head>
<body>
    ${content}
</body>
</html>`;
}

async function generatePDF(markdownPath, outputPath) {
    console.log(`📄 Procesando: ${path.basename(markdownPath)}`);

    // Leer el archivo markdown
    const markdown = fs.readFileSync(markdownPath, 'utf-8');

    // Convertir a HTML
    const htmlContent = markdownToHtml(markdown);

    // Crear HTML completo con estilos
    const cssPath = path.join(path.dirname(markdownPath), 'pdf-style.css');
    const title = path.basename(markdownPath, '.md');
    const fullHtml = createHtmlTemplate(title, htmlContent, cssPath);

    // Generar PDF con Playwright
    const browser = await chromium.launch();
    const page = await browser.newPage();

    await page.setContent(fullHtml, { waitUntil: 'networkidle' });

    await page.pdf({
        path: outputPath,
        format: 'A4',
        margin: {
            top: '2cm',
            right: '2cm',
            bottom: '2cm',
            left: '2cm'
        },
        printBackground: true,
        displayHeaderFooter: true,
        headerTemplate: `
            <div style="font-size: 9pt; color: #999; width: 100%; text-align: right; padding-right: 1cm;">
                Curso Intensivo de Español - CILE
            </div>
        `,
        footerTemplate: `
            <div style="font-size: 9pt; color: #999; width: 100%; text-align: center;">
                Página <span class="pageNumber"></span> de <span class="totalPages"></span>
            </div>
        `
    });

    await browser.close();

    console.log(`✅ PDF generado: ${path.basename(outputPath)}`);
}

async function main() {
    const materialesDir = __dirname;
    const pdfDir = path.join(materialesDir, 'pdfs');

    // Crear carpeta para PDFs si no existe
    if (!fs.existsSync(pdfDir)) {
        fs.mkdirSync(pdfDir, { recursive: true });
    }

    // Buscar todos los archivos .md
    const mdFiles = fs.readdirSync(materialesDir)
        .filter(file => file.endsWith('.md'))
        .map(file => path.join(materialesDir, file));

    console.log(`\n🚀 Generando PDFs para ${mdFiles.length} archivos...\n`);

    // Generar PDFs
    for (const mdFile of mdFiles) {
        const outputPath = path.join(
            pdfDir,
            path.basename(mdFile, '.md') + '.pdf'
        );

        try {
            await generatePDF(mdFile, outputPath);
        } catch (error) {
            console.error(`❌ Error procesando ${path.basename(mdFile)}:`, error.message);
        }
    }

    console.log('\n✨ ¡Proceso completado!\n');
}

// Ejecutar si es llamado directamente
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { generatePDF, markdownToHtml };
